src/main/scala/tile/BaseTile.scala

```diff
@@ -66,10 +65,22 @@ trait HasNonDiplomaticTileParameters {
     res
   }
   def asIdBits: Int = p(ASIdBits)
-  def maxPAddrBits: Int = xLen match { case 32 => 34; case 64 => 56 }
+  lazy val maxPAddrBits: Int = {
+    require(xLen == 32 || xLen == 64, s"We only support XLENs of 32 or 64, but got $xLen")
+    xLen match { case 32 => 34; case 64 => 56 }
+  }
 
-  def hartId: Int = tileParams.hartId
-  def hartIdLen: Int = p(MaxHartIdBits)
+  // Use staticIdForMetdata only to emit build information or identify a component to diplomacy.
+  //   Including it in a constructed Chisel circuit will prevent us
+  //   from being able to deduplicate otherwise-homogeneous tiles.
+  def staticIdForMetadata: Int = tileParams.hartId
+  @deprecated("use hartIdSinkNode.bundle or staticIdForMetadata", "rocket-chip 1.3")
+  def hartId: Int = staticIdForMetadata
+  lazy val hartIdLen: Int = {
+    val len = p(MaxHartIdBits)
+    require (log2Up(staticIdForMetadata + 1) <= len, s"p(MaxHartIdBits) of $len is not enough for static hart id $staticIdForMetadata")
```

Comment:
im a little confused by this requirement, given that staticIdForMetadata is only supposed to be used for metadata. But it still seems like it's getting used in the hardware somewhere. See my previous request for ASCII art for this :)

Reply:
doesn't p(MaxHartIdBits) also get used in $mhartid CSR? Well I guess i can RTFC to figure out how the width of that CSR is set, I just feel like i stumbled over this in the past

Reply:
ok, yeah, so currently CSR.scala uses `hartIdLen` to figure out how wide to make the mhartid CSR. Where is that coming from now...?

Reply:
https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/rocket/CSR.scala#L196

Reply:
basically I am saying the important thing is whatever CSR.scala is using to set that width of the `mhartid` CSR needs to be checking the actual width of whatever I wanted shown in my ASCII art diagram. That's what we want this require to actually be checking, I believe.

Reply:
The CSR's parent Core also has an IO sized by `hartIdLen`, i added a requirement checking the widths using Chisels `getWidth`